packages/contracts-bedrock/scripts/autogen/generate-snapshots.ts

```diff
@@ -89,23 +89,23 @@ const main = async () => {
 
   for (const contractFile of contractSources) {
     const contractArtifacts = path.join(forgeArtifactsDir, contractFile)
+    // Newer versions of forge omits artifacts for "libraries" that contain only error definitions, ex: CannonErrors.sol
+    if (!fs.existsSync(contractArtifacts)) {
+      console.log(`ignoring zero artifact contract ${contractFile}`)
+      continue
+    }
     for (const name of fs.readdirSync(contractArtifacts)) {
       const data = fs.readFileSync(path.join(contractArtifacts, name))
       const artifact: ForgeArtifact = JSON.parse(data.toString())
 
       const contractName = parseArtifactName(name)
 
-      // HACK: This is a hack to ignore libraries and abstract contracts. Not robust against changes to solc's internal ast repr
-      const isContract = artifact.ast.nodes.some((node: any) => {
-        return (
-          node.nodeType === 'ContractDefinition' &&
-          node.name === contractName &&
-          node.contractKind === 'contract' &&
-          (node.abstract === undefined || // solc < 0.6 doesn't have explicit abstract contracts
-            node.abstract === false)
-        )
-      })
-      if (!isContract) {
+      const isLibrary =
+        artifact.abi.length === 0 &&
+        (artifact.storageLayout === undefined ||
+          artifact.storageLayout.storage.length === 0)
+      const isAbstract = artifact.bytecode.object === '0x'
+      if (isLibrary || isAbstract) {
```

Comment:
Can we get away with setting `ast = true` in the `foundry.toml` and making no changes here? Seems like these changes were prompted by the removal of AST by default in the artifacts?

Reply:
Ahh I hoped such an option exists, but I couldn't find it in the foundry book or source code.
The new approach is I think less hacky - doesn't rely on ast internals. Though it might not be as robust.

Confirmed that `ast = true` retains this info. It seems to be an undocumented foundry feature so I'm a bit hesistant to rely on it.

Reply:
Found it sorta documented [here](https://github.com/foundry-rs/foundry/blob/f73d855ff32e94190c8c8b637ae47a73461efdab/crates/config/src/lib.rs#L326). Foundry just has poor documentation in general, but it seems to be a supported option.

I'll revert the script in favor of `ast = true`.

Reply:
Nit, but you could instead modify the snapshot command to build with `forge build --ast`. AST (de)serializing takes about ~1s in our repo (see benchmarks [here](https://github.com/foundry-rs/foundry/pull/7197)) so this would give a performance boost to non-snapshot builds, whereas the config file approach adds ~1s to all builds

Reply:
I think this makes sense to do.
Unfortunately, `--ast` isn't incremental. It takes just as long to build the ast with `--ast` as without on a warm build.
It could be annoying for users that use `pnpm pre-pr:no-build`. OTOH, snapshots aren't updated often, so overall it should save some time.